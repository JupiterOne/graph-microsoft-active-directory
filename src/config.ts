import {
  IntegrationExecutionContext,
  IntegrationValidationError,
  IntegrationInstanceConfigFieldMap,
  IntegrationInstanceConfig,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from './client';

/**
 * Config fields for the Microsoft Active Directory (on-prem version).
 */
export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  adUsername: {
    type: 'string',
  },
  adPassword: {
    type: 'string',
    mask: true,
  },
  ldapUrl: {
    type: 'string',
  },
  baseDN: {
    type: 'string',
  },
};

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * The URL of the provider LDAP server.
   * ie - ldap://127.0.0.1
   */
  ldapUrl: string;

  /**
   * The Active Directory base Distinguished Name.
   * This is used as the entry point for searches.
   * ie - dc=activedir,dc=com
   */
  baseDN: string;

  /**
   * An Active Directory username.
   * NOTE: this cannot be renamed to username.
   *       Env var USERNAME not usable on Windows
   */
  adUsername: string;

  /**
   * The Active Directory password.
   */
  adPassword: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  const { config } = context.instance;

  if (
    !config.ldapUrl ||
    !config.baseDN ||
    !config.adUsername ||
    !config.adPassword
  ) {
    throw new IntegrationValidationError(
      'Config requires all of {ldapUrl, baseDN, adUsername, adPassword}',
    );
  }

  const apiClient = createAPIClient(config);
  await apiClient.verifyAuthentication();
}
