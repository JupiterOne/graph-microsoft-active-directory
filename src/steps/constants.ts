import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export const Steps = {
  ACCOUNT: 'fetch-account',
  USERS: 'fetch-users',
  GROUPS: 'fetch-groups',
  DEVICES: 'fetch-devices',
  GROUP_USER_RELATIONSHIPS: 'build-user-group-relationships',
  GROUP_GROUP_RELATIONSHIPS: 'build-group-group-relationships',
};

export const Entities: Record<
  'ACCOUNT' | 'GROUP' | 'USER' | 'DEVICE',
  StepEntityMetadata
> = {
  ACCOUNT: {
    resourceName: 'Account',
    _type: 'ad_account',
    _class: ['Account'],
    schema: {
      properties: {},
      required: [],
    },
  },
  GROUP: {
    resourceName: 'UserGroup',
    _type: 'ad_group',
    _class: ['UserGroup'],
    schema: {
      properties: {
        name: { type: 'string' },
        description: { type: 'string' },
        createdOn: { type: 'number' },
        updatedOn: { type: 'number' },
      },
      required: ['name', 'description', 'createdOn', 'updatedOn'],
    },
  },
  USER: {
    resourceName: 'User',
    _type: 'ad_user',
    _class: ['User'],
    schema: {
      properties: {
        username: { type: 'string' },
        active: { type: 'boolean' },
        description: { type: 'string' },
        createdOn: { type: 'number' },
        updatedOn: { type: 'number' },
      },
      required: ['username', 'active', 'description', 'createdOn', 'updatedOn'],
    },
  },
  DEVICE: {
    resourceName: 'Device',
    _type: 'ad_device',
    _class: ['Device'],
    schema: {
      properties: {
        operatingSystem: { type: 'string' },
        operatingSystemVersion: { type: 'string' },
        createdOn: { type: 'number' },
        updatedOn: { type: 'number' },
      },
      required: [
        'operatingSystem',
        'operatingSystemVersion',
        'createdOn',
        'updatedOn',
      ],
    },
  },
};

export const Relationships: Record<
  | 'ACCOUNT_HAS_USER'
  | 'ACCOUNT_HAS_GROUP'
  | 'ACCOUNT_HAS_DEVICE'
  | 'GROUP_HAS_USER'
  | 'GROUP_HAS_GROUP',
  StepRelationshipMetadata
> = {
  ACCOUNT_HAS_USER: {
    _type: 'ad_account_has_user',
    sourceType: Entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.USER._type,
  },
  ACCOUNT_HAS_GROUP: {
    _type: 'ad_account_has_group',
    sourceType: Entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.GROUP._type,
  },
  ACCOUNT_HAS_DEVICE: {
    _type: 'ad_account_has_device',
    sourceType: Entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.DEVICE._type,
  },
  GROUP_HAS_USER: {
    _type: 'ad_group_has_user',
    sourceType: Entities.GROUP._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.USER._type,
  },
  GROUP_HAS_GROUP: {
    _type: 'ad_group_has_group',
    sourceType: Entities.GROUP._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.GROUP._type,
  },
};
