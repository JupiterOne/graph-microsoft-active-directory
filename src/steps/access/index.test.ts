import {
  executeStepWithDependencies,
  Recording,
  setupRecording,
} from '@jupiterone/integration-sdk-testing';
import { buildStepTestConfigForStep } from '../../../test/config';
import { Steps } from '../constants';

let recording: Recording;
afterEach(async () => {
  await recording.stop();
});

describe('access steps', () => {
  test('fetch-users', async () => {
    recording = setupRecording({
      directory: __dirname,
      name: 'fetch-users',
    });
    const stepConfig = buildStepTestConfigForStep(Steps.USERS);
    const stepResult = await executeStepWithDependencies(stepConfig);

    expect(stepResult).toMatchStepMetadata(stepConfig);
  });

  test('fetch-groups', async () => {
    const stepConfig = buildStepTestConfigForStep(Steps.GROUPS);
    const stepResult = await executeStepWithDependencies(stepConfig);
    expect(stepResult).toMatchStepMetadata(stepConfig);
  });

  test('build-user-group-relationships', async () => {
    const stepConfig = buildStepTestConfigForStep(
      Steps.GROUP_USER_RELATIONSHIPS,
    );
    const stepResult = await executeStepWithDependencies(stepConfig);
    expect(stepResult).toMatchStepMetadata(stepConfig);
  });
});
