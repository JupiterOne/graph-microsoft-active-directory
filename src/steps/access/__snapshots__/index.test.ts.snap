// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build-user-group-relationships 1`] = `
Object {
  "_class": "HAS",
  "_fromEntityKey": "ad_group:CN=Group Policy Creator Owners,CN=Users,DC=activedirint,DC=com",
  "_key": "ad_group:CN=Group Policy Creator Owners,CN=Users,DC=activedirint,DC=com|has|ad_user:CN=activedir,CN=Users,DC=activedirint,DC=com",
  "_toEntityKey": "ad_user:CN=activedir,CN=Users,DC=activedirint,DC=com",
  "_type": "ad_group_has_user",
  "displayName": "HAS",
}
`;

exports[`fetch-devices 1`] = `
Object {
  "_class": Array [
    "Device",
  ],
  "_key": "ad_device:CN=ActiveDirectory,OU=Domain Controllers,DC=activedirint,DC=com",
  "_type": "ad_device",
  "category": "computer",
  "createdOn": 1647041559000,
  "deviceId": "CN=ActiveDirectory,OU=Domain Controllers,DC=activedirint,DC=com",
  "displayName": "ActiveDirectory",
  "isCriticalSystemObject": true,
  "lastLogon": undefined,
  "lastSeenOn": Any<Number>,
  "make": null,
  "model": null,
  "name": "ActiveDirectory",
  "operatingSystem": "Windows Server 2019 Datacenter",
  "operatingSystemVersion": "10.0 (17763)",
  "serial": null,
  "updatedOn": 1649751405000,
}
`;

exports[`fetch-groups 1`] = `
Object {
  "_class": Array [
    "UserGroup",
  ],
  "_key": "ad_group:CN=Administrators,CN=Builtin,DC=activedirint,DC=com",
  "_type": "ad_group",
  "createdOn": 1647041514000,
  "description": "Administrators have complete and unrestricted access to the computer/domain",
  "displayName": "Administrators",
  "name": "Administrators",
  "updatedOn": 1647042471000,
}
`;

exports[`fetch-users 1`] = `
Object {
  "_class": Array [
    "User",
  ],
  "_key": "ad_user:CN=activedir,CN=Users,DC=activedirint,DC=com",
  "_type": "ad_user",
  "active": true,
  "createdOn": 1647041514000,
  "description": "Built-in account for administering the computer/domain",
  "displayName": "activedir",
  "name": "activedir",
  "updatedOn": 1650392064000,
  "username": "activedir",
}
`;
